moj_utils/healthchecks.py

```diff
@@ -0,0 +1,205 @@
+                                               json_path)
+                if value != expected_value:
+                    return self.error_response('Response JSON path "%s" did not contain "%s"' %
+                                               (json_path, expected_value))
+        except requests.Timeout:
+            return self.error_response('Timed out')
+        except requests.TooManyRedirects:
+            return self.error_response('URL redirected too many times')
+        except requests.HTTPError:
+            return self.error_response('URL not loaded')
+
+        return HealthcheckResponse(self.name, True, url=self.url)
+
+    def error_response(self, error):
+        return HealthcheckResponse(self.name, False, error=error,
+                                   url=self.url)
+
+
+class HealthcheckRegistry(object):
+    """
+    Healthcheck registry - loads and runs healthchecks.  A healthcheck is a
+    callable, or a class whose instances are callable, that returns a bool,
+    a HealthcheckResponse or raises an exception.
+    """
+
+    def __init__(self):
+        self._registry = []
+        self._registry_loaded = False
+
+    def reset(self):
+        self._registry = []
+        self._registry_loaded = False
+
+    def load_healthchecks(self):
+        """
+        Loads healthchecks.
+        """
+        self.load_default_healthchecks()
+        if getattr(settings, 'AUTODISCOVER_HEALTHCHECKS', True):
+            self.autodiscover_healthchecks()
+        self._registry_loaded = True
+
+    def load_default_healthchecks(self):
+        """
+        Loads healthchecks specified in settings.HEALTHCHECKS as dotted import
+        paths to the classes. Defaults are listed in `DEFAULT_HEALTHCHECKS`.
+        """
+        default_healthchecks = getattr(settings, 'HEALTHCHECKS', DEFAULT_HEALTHCHECKS)
+        for healthcheck in default_healthchecks:
+            healthcheck = import_string(healthcheck)
+            self.register_healthcheck(healthcheck)
+
+    def autodiscover_healthchecks(self):
+        """
+        Loads healthchecks.py in all installed apps. Register healthchecks
+        in your healthchecks.py by calling `registry.register_healthcheck`
+        """
+        autodiscover_modules('healthchecks', register_to=self)
+
+    def register_healthcheck(self, healthcheck):
+        """
+        Register a healthcheck. Call this method from your apps'
+        healthchecks.py and use autodiscovery for loading.
+        :param healthcheck: callable or callable class
+        """
+        self._registry.append(healthcheck)
+
+    def run_healthchecks(self):
+        """
+        Runs all registered healthchecks and returns a list of
+        HealthcheckResponse.
+        """
+        if not self._registry_loaded:
+            self.load_healthchecks()
+
+        def get_healthcheck_name():
+            if hasattr(healthcheck, 'name'):
+                return healthcheck.name
+            return healthcheck.__name__
```

Comment:
I don't like using variables from the scope of the containing function in an inner function - I prefer to see any required variables passed in as argument. That might just be me, though.